### DISCUSSION TITLE
Model library

### DISCUSSION THREAD

**Discussion by timgjohnson** [talmolab/sleap#1437] [Ideas]
Hi,

It could be very helpful to have a library of well-trained models that people could download and use in their projects. Perhaps there could even be a place for people to upload their well-performing models so that others could use them. It seems that it would be beneficial for everyone if more people could use high-performing models to the extent that they can generalize. 

Thank you!

**Comment by talmo** [comment]
Hi @timgjohnson,

Fully agreed!! We're planning on developing several web-based components to support use cases surrounding data and model sharing workflows over the next year.

For something like model sharing there are some technical and conceptual hurdles we still have to sort out.

For example, models by themselves don't provide much context in terms of what kind of data it works well on, so to make it practical you probably need some metadata on the dataset it was trained on. In the ideal case, we'd like to enforce some standards such as [model cards](https://arxiv.org/abs/1810.03993) together with each entry in the library. This adds a lot of burden on the user side to "document" their model, which might hamper usage of such a feature. (What's in it for the uploader if they have to do all this work?)

Additional metadata that would be important to include would be performance metrics, but doing appropriate evaluation is hard since it requires train/val/test splits and replicates, not to mention the choice of appropriate metrics to report.

As a SLEAP-specific concern: how do we treat multi-stage models like top-down models that have two independently trainable models (centroid + centered instance)? Should they be treated as separate entries in such a library?

Finally, what do we do about intellectual ownership and attribution? If others are going to reuse models, then it should be trackable and reliably identifiable (and maybe versioned?) to ensure reproducibility. We'd like to generate RRIDs for each model, but integrating this might take some serious work. In terms of IP rights, who owns the model weights?

In any case, we should probably tackle these issues one step at a time, but if we're going to do it we'd like to do it ethically and responsibly. We'd love input if you have any thoughts on any of these concerns, and thanks for posting!

Cheers,

Talmo
